### ISSUE TITLE
tonic-tracing-opentelemetry 0.15: spans being marked with errors even though everything looks fine otherwise.

### ISSUE THREAD

**Issue by nshalman** [davidB/tracing-opentelemetry-instrumentation-sdk#122]
Example of a surprising error:
```
     0.003207564s TRACE otel::tracing: close, time.busy: 54.0µs, time.idle: 2.47ms
    at /home/nshalman/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-opentelemetry-instrumentation-sdk-0.15.0/src/http/grpc_client.rs:13 on main
    in otel::tracing::GRPC request with http.user_agent: , otel.name: "helloworld.Greeter/SayHello", otel.kind: Client, rpc.system: "grpc", rpc.service: helloworld.Greeter, rpc.method: SayHello, server.address: , rpc.grpc.status_code: 2, otel.status_code: "ERROR"
```
I don't understand why the spans are being marked as errors.

Changes I made to get the example code to compile:

```diff
diff --git a/examples/grpc/Cargo.toml b/examples/grpc/Cargo.toml
index a8884a3..6076d9e 100644
--- a/examples/grpc/Cargo.toml
+++ b/examples/grpc/Cargo.toml
@@ -19,10 +19,10 @@ tonic = { version = "0.10", features = [
 ], default-features = false }
 tonic-health = "0.10"
 tonic-reflection = "0.10"
-tonic-tracing-opentelemetry = { path = "../../tonic-tracing-opentelemetry" }
+tonic-tracing-opentelemetry = "0.15"
 tower = "0.4"
 tracing = "0.1"
-tracing-opentelemetry-instrumentation-sdk = { path = "../../tracing-opentelemetry-instrumentation-sdk" }
+tracing-opentelemetry-instrumentation-sdk = "0.15"

 [build-dependencies]
 tonic-build = { version = "0.10", default-features = false, features = [
```

In `examples/grpc` run (pointed at a local jaeger):

```
export OTEL_EXPORTER_OTLP_ENDPOINT=http://localhost:4317
export OTEL_SERVICE_NAME=server; cargo run --bin examples-grpc-server &
export OTEL_SERVICE_NAME=client; cargo run --bin client
```

Output (ignoring startup messages):
```
     0.000683734s TRACE otel::tracing: new
    at /home/nshalman/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-opentelemetry-instrumentation-sdk-0.15.0/src/http/grpc_client.rs:13 on main
    in otel::tracing::GRPC request with http.user_agent: , otel.name: "helloworld.Greeter/SayHello", otel.kind: Client, rpc.system: "grpc", rpc.service: helloworld.Greeter, rpc.method: SayHello, server.address:

    64.031057174s TRACE otel::tracing: new
    at /home/nshalman/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-opentelemetry-instrumentation-sdk-0.15.0/src/http/grpc_server.rs:11 on tokio-runtime-worker
    in otel::tracing::GRPC request with http.user_agent: tonic/0.10.2, otel.name: "helloworld.Greeter/SayHello", otel.kind: Server, rpc.system: "grpc", rpc.service: helloworld.Greeter, rpc.method: SayHello, server.address: 127.0.0.1

    64.031411009s  INFO server: new
    at src/server.rs:18 on tokio-runtime-worker
    in server::say_hello
    in otel::tracing::GRPC request with http.user_agent: tonic/0.10.2, otel.name: "helloworld.Greeter/SayHello", otel.kind: Server, rpc.system: "grpc", rpc.service: helloworld.Greeter, rpc.method: SayHello, server.address: 127.0.0.1

Got a request from Some(127.0.0.1:44644) (Some("6e51442f8cacbc01252a8ceb391f8b13"))
    64.031509446s  INFO server: close, time.busy: 41.0µs, time.idle: 59.4µs
    at src/server.rs:18 on tokio-runtime-worker
    in server::say_hello
    in otel::tracing::GRPC request with http.user_agent: tonic/0.10.2, otel.name: "helloworld.Greeter/SayHello", otel.kind: Server, rpc.system: "grpc", rpc.service: helloworld.Greeter, rpc.method: SayHello, server.address: 127.0.0.1

    64.031676665s TRACE otel::tracing: close, time.busy: 316µs, time.idle: 309µs
    at /home/nshalman/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-opentelemetry-instrumentation-sdk-0.15.0/src/http/grpc_server.rs:11 on tokio-runtime-worker
    in otel::tracing::GRPC request with http.user_agent: tonic/0.10.2, otel.name: "helloworld.Greeter/SayHello", otel.kind: Server, rpc.system: "grpc", rpc.service: helloworld.Greeter, rpc.method: SayHello, server.address: 127.0.0.1, rpc.grpc.status_code: 2, otel.status_code: "ERROR"

     0.003207564s TRACE otel::tracing: close, time.busy: 54.0µs, time.idle: 2.47ms
    at /home/nshalman/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-opentelemetry-instrumentation-sdk-0.15.0/src/http/grpc_client.rs:13 on main
    in otel::tracing::GRPC request with http.user_agent: , otel.name: "helloworld.Greeter/SayHello", otel.kind: Client, rpc.system: "grpc", rpc.service: helloworld.Greeter, rpc.method: SayHello, server.address: , rpc.grpc.status_code: 2, otel.status_code: "ERROR"

RESPONSE=Response { metadata: MetadataMap { headers: {"content-type": "application/grpc", "date": "Fri, 12 Jan 2024 20:59:09 GMT", "grpc-status": "0"} }, message: HelloReply { message: "Hello Tonic! (Some(\"6e51442f8cacbc01252a8ceb391f8b13\"))" }, extensions: Extensions }
```

**Comment by nshalman** [comment]
I had to go all the way back to tag "0.13.1" to get the expected output (specifically that the status code is "OK"):
```
     0.002136207s TRACE otel::tracing: new
    at tracing-opentelemetry-instrumentation-sdk/src/http/grpc_client.rs:11 on main
    in otel::tracing::GRPC request with http.user_agent: , otel.name: "helloworld.Greeter/SayHello", otel.kind: Client, rpc.system: "grpc", rpc.service: helloworld.Greeter, rpc.method: SayHello, server.address: 

     7.089852930s TRACE otel::tracing: new
    at tracing-opentelemetry-instrumentation-sdk/src/http/grpc_server.rs:9 on tokio-runtime-worker
    in otel::tracing::GRPC request with http.user_agent: tonic/0.9.2, otel.name: "helloworld.Greeter/SayHello", otel.kind: Server, rpc.system: "grpc", rpc.service: helloworld.Greeter, rpc.method: SayHello, server.address: 127.0.0.1

     7.090255011s  INFO server: new
    at examples/grpc/src/server.rs:18 on tokio-runtime-worker
    in server::say_hello
    in otel::tracing::GRPC request with http.user_agent: tonic/0.9.2, otel.name: "helloworld.Greeter/SayHello", otel.kind: Server, rpc.system: "grpc", rpc.service: helloworld.Greeter, rpc.method: SayHello, server.address: 127.0.0.1

Got a request from Some(127.0.0.1:36654) (Some("8053b6e010a73a2227e2947076732324"))
     7.090378298s  INFO server: close, time.busy: 52.6µs, time.idle: 73.4µs
    at examples/grpc/src/server.rs:18 on tokio-runtime-worker
    in server::say_hello
    in otel::tracing::GRPC request with http.user_agent: tonic/0.9.2, otel.name: "helloworld.Greeter/SayHello", otel.kind: Server, rpc.system: "grpc", rpc.service: helloworld.Greeter, rpc.method: SayHello, server.address: 127.0.0.1

     7.090625452s TRACE otel::tracing: close, time.busy: 443µs, time.idle: 342µs
    at tracing-opentelemetry-instrumentation-sdk/src/http/grpc_server.rs:9 on tokio-runtime-worker
    in otel::tracing::GRPC request with http.user_agent: tonic/0.9.2, otel.name: "helloworld.Greeter/SayHello", otel.kind: Server, rpc.system: "grpc", rpc.service: helloworld.Greeter, rpc.method: SayHello, server.address: 127.0.0.1, http.status_code: 200, otel.status_code: "OK"

     0.005856001s TRACE otel::tracing: close, time.busy: 81.9µs, time.idle: 3.65ms
    at tracing-opentelemetry-instrumentation-sdk/src/http/grpc_client.rs:11 on main
    in otel::tracing::GRPC request with http.user_agent: , otel.name: "helloworld.Greeter/SayHello", otel.kind: Client, rpc.system: "grpc", rpc.service: helloworld.Greeter, rpc.method: SayHello, server.address: , http.status_code: 200, otel.status_code: "OK"
```

**Comment by davidB** [comment]
iirc before 0.14, the trace'status for grpc call was wrong because based on HTTP status and not on grpc status.

The first call you liked has a grpc status == 2 (UNKNOWN). So the bug, IMHO, is why the grpc status is `UNKNOWN` for the sample hello world?

ref:

- https://github.com/grpc/grpc/blob/master/doc/statuscodes.md
- https://opentelemetry.io/docs/specs/semconv/rpc/grpc/#grpc-status

**Comment by nshalman** [comment]
> iirc before 0.14, the trace'status for grpc call was wrong because based on HTTP status and not on grpc status.
> 
> The first call you liked has a grpc status == 2 (UNKNOWN). So the bug, IMHO, is why the grpc status is `UNKNOWN` for the sample hello world?
> 
> ref:
> 
> * https://github.com/grpc/grpc/blob/master/doc/statuscodes.md
> * https://opentelemetry.io/docs/specs/semconv/rpc/grpc/#grpc-status

So interestingly I still see the those errors on the rust client side when run against a golang server where there is no error on the server side of the RPC AND the response is fully parsed out successfully. I do think there is a bug somewhere that is incorrectly marking the RPC as an error just in the span when it is not actually an error anywhere else. I'm still new to rust so I'm not quite sure how to prove it yet.

**Comment by davidB** [comment]
I'll add some test on https://github.com/davidB/tracing-opentelemetry-instrumentation-sdk/blob/d3609ac2cc699d3a24fbf89754053cc8e938e3bf/tracing-opentelemetry-instrumentation-sdk/src/http/tools.rs#L130 but also on the way the grpc status is read (it's probably where is the bug)

**Comment by davidB** [comment]
I noticed strange behavior during my quick try, no grpc status available at layer level (client and server). Fix will require deeper investigation inside tonic.

For note to myself & illustration (with debug on client):

```
[tracing-opentelemetry-instrumentation-sdk/src/http/tools.rs:182] headers = {
    "content-type": "application/grpc",
    "date": "Sun, 21 Jan 2024 15:28:55 GMT",
}
[tracing-opentelemetry-instrumentation-sdk/src/http/tools.rs:168] grpc_status_from_http_header(response.headers()) = None
[tracing-opentelemetry-instrumentation-sdk/src/http/tools.rs:169] grpc_status_from_http_status(response.status()) = None
     0.002579164s TRACE otel::tracing: close, time.busy: 147µs, time.idle: 1.86ms
    at tracing-opentelemetry-instrumentation-sdk/src/http/grpc_client.rs:13 on main
    in otel::tracing::GRPC request with http.user_agent: , otel.name: "helloworld.Greeter/SayHello", otel.kind: Client, rpc.system: "grpc", rpc.service: helloworld.Greeter, rpc.method: SayHello, server.address: , rpc.grpc.status_code: 2, otel.status_code: "ERROR"

RESPONSE=Response { metadata: MetadataMap { headers: {"content-type": "application/grpc", "date": "Sun, 21 Jan 2024 15:28:55 GMT", "grpc-status": "0"} }, message: HelloReply { message: "Hello Tonic! (Some(\"417b58baeecc78e77a8575a3bc3809e1\"))" }, extensions: Extensions }
```

`dbg` calls are done at layer level, `RESPONSE=...` is done outside of tonic, after the client received the full response.

**Comment by davidB** [comment]
Fixed with 0.15.2

**Comment by nshalman** [comment]
Thank you!
